### PR TITLE
feat(VcNarrator): add `{{DISPLAY_NAME}}` as placeholder

### PR DESCRIPTION
**Pull request by Marocco2**
`{{DISPLAY_NAME}}` is placeholder for user's `global_name` object if it exists, otherwise it defaults to `username`

### CODE REVIEW

File: src/plugins/vcNarrator.tsx [hunk 1]
```diff
@@ -143,7 +144,7 @@ function updateStatuses(type: string, { deaf, mute, selfDeaf, selfMute, userId,
 function playSample(tempSettings: any, type: string) {
     const settings = Object.assign({}, Settings.plugins.VcNarrator, tempSettings);
 
-    speak(formatText(settings[type + "Message"], UserStore.getCurrentUser().username, "general"), settings);
+    speak(formatText(settings[type + "Message"], UserStore.getCurrentUser().username, "general", (UserStore.getUser(userId).global_name ? UserStore.getUser(userId).global_name : UserStore.getUser(userId).username)), settings);
```
Comment (TheKodeToad) [inline src/plugins/vcNarrator.tsx:147]:
```suggestion
    speak(formatText(settings[type + "Message"], UserStore.getCurrentUser().username, "general", UserStore.getUser(userId).global_name ?? UserStore.getUser(userId).username), settings);
```

**Comment by Marocco2** [comment]
Any ETAs?

**Comment by Marocco2** [comment]
Alright, gonna check the code

File: src/plugins/vcNarrator.tsx [hunk 2]
```diff
@@ -171,9 +172,10 @@ export default definePlugin({
 
                 const template = Settings.plugins.VcNarrator[type + "Message"];
                 const user = isMe && !Settings.plugins.VcNarrator.sayOwnName ? "" : UserStore.getUser(userId).username;
+                const displayNameUser = isMe && !Settings.plugins.VcNarrator.sayOwnName ? "" : UserStore.getUser(userId).globalName ?? UserStore.getUser(userId).username;
```
Comment (Marocco2) [inline src/plugins/vcNarrator.tsx:175]: Camel-case, my nemesis...

**Comment by Marocco2** [comment]
Now it should compile

File: src/plugins/vcNarrator.tsx [hunk 3]
```diff
@@ -311,8 +313,8 @@ export default definePlugin({
                     You can customise the spoken messages below. You can disable specific messages by setting them to nothing
                 </Forms.FormText>
                 <Forms.FormText>
-                    The special placeholders <code>{"{{USER}}"}</code> and <code>{"{{CHANNEL}}"}</code>{" "}
-                    will be replaced with the user's name (nothing if it's yourself) and the channel's name respectively
+                    The special placeholders <code>{"{{USER}}"}</code>, <code>{"{{DISPLAY_NAME_USER}}"}</code> and <code>{"{{CHANNEL}}"}</code>{" "}
+                    will be replaced with the user's name (nothing if it's yourself), user's displayed name and the channel's name respectively
```
Comment (Vendicated) [inline src/plugins/vcNarrator.tsx:317]:
```suggestion
                    will be replaced with the user's name (nothing if it's yourself), the user's display name and the channel's name respectively
```

File: src/plugins/vcNarrator.tsx [hunk 4]
```diff
@@ -69,10 +69,11 @@ function clean(str: string) {
         .trim();
 }
 
-function formatText(str: string, user: string, channel: string) {
+function formatText(str: string, user: string, channel: string, displayNameUser: string) {
```
Comment (Vendicated) [inline src/plugins/vcNarrator.tsx:72]: displayName

File: src/plugins/vcNarrator.tsx [hunk 5]
```diff
@@ -69,10 +69,11 @@ function clean(str: string) {
...
     return str
         .replaceAll("{{USER}}", clean(user) || (user ? "Someone" : ""))
-        .replaceAll("{{CHANNEL}}", clean(channel) || "channel");
+        .replaceAll("{{CHANNEL}}", clean(channel) || "channel")
+        .replaceAll("{{DISPLAY_NAME_USER}}", clean(displayNameUser) || (displayNameUser ? "Someone" : ""));
```
Comment (Vendicated) [inline src/plugins/vcNarrator.tsx:76]:
```suggestion
        .replaceAll("{{DISPLAY_NAME}}", clean(displayNameUser) || (displayNameUser ? "Someone" : ""));
```

**Comment by Marocco2** [comment]
@Vendicated Changes done 👍

File: src/plugins/vcNarrator/index.tsx [hunk 6]
```diff
@@ -172,9 +173,10 @@ export default definePlugin({
 
                 const template = Settings.plugins.VcNarrator[type + "Message"];
                 const user = isMe && !Settings.plugins.VcNarrator.sayOwnName ? "" : UserStore.getUser(userId).username;
+                const displayName = isMe && !Settings.plugins.VcNarrator.sayOwnName ? "" : UserStore.getUser(userId).globalName ?? UserStore.getUser(userId).username;
```
Comment (Vendicated) [inline src/plugins/vcNarrator/index.tsx:176]:
```suggestion
                const displayName = user && (UserStore.getUser(userId).globalName ?? user);
```

**Comment by Marocco2** [comment]
Is it gonna be green lit?

**Comment by Marocco2** [comment]
What prevents this to be compiled is https://github.com/Swishilicous/discord-types being not updated on NPM while having that new `globalName` property: https://github.com/Swishilicous/discord-types/pull/2

I would suggest the following solutions:

- Press `discord-types` maintainer to update those deps on NPM
- Fork it

@Vendicated sorry for this inconvenience

**Comment by TheKodeToad** [comment]
Just use `as any` like in ShowMeYourName :duck:

**Comment by Marocco2** [comment]
> Just use `as any` like in ShowMeYourName :duck: 

Thanks, gonna try that 👍

**Comment by ImLvna** [comment]
> Just use `as any` like in ShowMeYourName 🦆

please use `as string` or `as any as string` to keep type safety :fear:

File: src/plugins/vcNarrator/index.tsx [hunk 7]
```diff
@@ -172,9 +173,10 @@ export default definePlugin({
 
                 const template = Settings.plugins.VcNarrator[type + "Message"];
                 const user = isMe && !Settings.plugins.VcNarrator.sayOwnName ? "" : UserStore.getUser(userId).username;
+                const displayName = user && ((UserStore.getUser(userId) as any as string).globalName ?? user);
```
Comment (Vendicated) [inline src/plugins/vcNarrator/index.tsx:176]:
```suggestion
                const displayName = user && ((UserStore.getUser(userId) as any).globalName ?? user);
```

File: src/plugins/vcNarrator/index.tsx [hunk 8]
```diff
@@ -144,7 +145,7 @@ function updateStatuses(type: string, { deaf, mute, selfDeaf, selfMute, userId,
 function playSample(tempSettings: any, type: string) {
     const settings = Object.assign({}, Settings.plugins.VcNarrator, tempSettings);
 
-    speak(formatText(settings[type + "Message"], UserStore.getCurrentUser().username, "general"), settings);
+    speak(formatText(settings[type + "Message"], UserStore.getCurrentUser().username, "general", (UserStore.getCurrentUser() as any as string).globalName ?? UserStore.getCurrentUser().username), settings);
```
Comment (Vendicated) [inline src/plugins/vcNarrator/index.tsx:148]: use a variable to make it cleaner, also `as any as string` is nonsense

**Comment by Vendicated** [comment]
thanks

**Pull request merged**